File: src/exonware/xwsystem/structures/contracts.py
#exonware/xwsystem/src/exonware/xwsystem/structures/contracts.py
"""
Company: eXonware.com
Author: Eng. Muhammad AlShehri
Email: connect@exonware.com
Version: 0.1.0.4
Generation Date: September 04, 2025

Structures module contracts - interfaces and enums for data structure functionality.
"""

from __future__ import annotations

from typing import Any, Optional, Iterator, Callable, Protocol, runtime_checkable

# Import enums from types module
from .defs import (
    StructureType,
    TraversalOrder,
    TraversalType,
    GraphType,
    CircularDetectionMethod,
    ValidationLevel
)


@runtime_checkable
class ITreeNode(Protocol):
    """Interface for tree node operations."""
    
    @property
    def value(self) -> Any:
        """Node value."""
        ...
    
    @property
    def children(self) -> list[ITreeNode]:
        """Node children."""
        ...
    
    @property
    def parent(self) -> Optional[ITreeNode]:
        """Node parent."""
        ...
    
    def add_child(self, child: ITreeNode) -> None:
        """Add child node."""
        ...
    
    def remove_child(self, child: ITreeNode) -> None:
        """Remove child node."""
        ...
    
    def is_leaf(self) -> bool:
        """Check if node is leaf."""
        ...
    
    def get_depth(self) -> int:
        """Get node depth."""
        ...


@runtime_checkable
class ITreeWalker(Protocol):
    """Interface for tree walking operations."""
    
    def walk_preorder(self, root: ITreeNode) -> Iterator[ITreeNode]:
        """Walk tree in preorder."""
        ...
    
    def walk_inorder(self, root: ITreeNode) -> Iterator[ITreeNode]:
        """Walk tree in inorder."""
        ...
    
    def walk_postorder(self, root: ITreeNode) -> Iterator[ITreeNode]:
        """Walk tree in postorder."""
        ...
    
    def walk_level_order(self, root: ITreeNode) -> Iterator[ITreeNode]:
        """Walk tree in level order."""
        ...
    
    def find_nodes(self, root: ITreeNode, predicate: Callable[[ITreeNode], bool]) -> list[ITreeNode]:
        """Find nodes matching predicate."""
        ...


@runtime_checkable
class ICircularDetector(Protocol):
    """Interface for circular reference detection."""
    
    def detect_circular_reference(self, obj: Any) -> bool:
        """Detect circular reference in object."""
        ...
    
    def find_circular_paths(self, obj: Any) -> list[list[Any]]:
        """Find all circular paths."""
        ...
    
    def get_circular_objects(self, obj: Any) -> list[Any]:
        """Get objects involved in circular references."""
        ...
    
    def break_circular_references(self, obj: Any) -> Any:
        """Break circular references."""
        ...


@runtime_checkable
class IGraphNode(Protocol):
    """Interface for graph node operations."""
    
    @property
    def id(self) -> str:
        """Node ID."""
        ...
    
    @property
    def data(self) -> Any:
        """Node data."""
        ...
    
    @property
    def neighbors(self) -> list[IGraphNode]:
        """Node neighbors."""
        ...
    
    def add_neighbor(self, neighbor: IGraphNode, weight: Optional[float] = None) -> None:
        """Add neighbor node."""
        ...
    
    def remove_neighbor(self, neighbor: IGraphNode) -> None:
        """Remove neighbor node."""
        ...
    
    def get_edge_weight(self, neighbor: IGraphNode) -> Optional[float]:
        """Get edge weight to neighbor."""
        ...


@runtime_checkable
class IGraph(Protocol):
    """Interface for graph operations."""
    
    def add_node(self, node: IGraphNode) -> None:
        """Add node to graph."""
        ...
    
    def remove_node(self, node_id: str) -> None:
        """Remove node from graph."""
        ...
    
    def get_node(self, node_id: str) -> Optional[IGraphNode]:
        """Get node by ID."""
        ...
    
    def add_edge(self, from_node: str, to_node: str, weight: Optional[float] = None) -> None:
        """Add edge between nodes."""
        ...
    
    def remove_edge(self, from_node: str, to_node: str) -> None:
        """Remove edge between nodes."""
        ...
    
    def get_all_nodes(self) -> list[IGraphNode]:
        """Get all nodes."""
        ...
    
    def get_all_edges(self) -> list[tuple]:
        """Get all edges."""
        ...
    
    def is_connected(self, from_node: str, to_node: str) -> bool:
        """Check if nodes are connected."""
        ...

File: rust/src/structures/contracts.rs
// #exonware/xwsystem/rust/src/structures/contracts.rs
//! Company: eXonware.com
//! Author: eXonware Backend Team
//! Email: connect@exonware.com
//! Version: 0.1.0.1
//! Generation Date: September 04, 2025
//! 
//! Structures module contracts - interfaces and enums for data structure functionality.


use crate::defs::{CircularDetectionMethod, GraphType, StructureType, TraversalOrder, TraversalType, ValidationLevel};

/// Interface for tree node operations.
pub trait ITreeNode {
    /// Node value.
    // Python decorators: @property
    fn value(&self) -> serde_json::Value;

    /// Node children.
    // Python decorators: @property
    fn children(&self) -> Vec<String>;

    /// Node parent.
    // Python decorators: @property
    fn parent(&self) -> Option<String>;

    /// Add child node.
    fn add_child(&self, child: String) -> ();

    /// Remove child node.
    fn remove_child(&self, child: String) -> ();

    /// Check if node is leaf.
    fn is_leaf(&self) -> bool;

    /// Get node depth.
    fn get_depth(&self) -> i64;

}

/// Interface for tree walking operations.
pub trait ITreeWalker {
    /// Walk tree in preorder.
    fn walk_preorder(&self, root: ITreeNode) -> Box<dyn Iterator<Item = ITreeNode>>;

    /// Walk tree in inorder.
    fn walk_inorder(&self, root: ITreeNode) -> Box<dyn Iterator<Item = ITreeNode>>;

    /// Walk tree in postorder.
    fn walk_postorder(&self, root: ITreeNode) -> Box<dyn Iterator<Item = ITreeNode>>;

    /// Walk tree in level order.
    fn walk_level_order(&self, root: ITreeNode) -> Box<dyn Iterator<Item = ITreeNode>>;

    /// Find nodes matching predicate.
    fn find_nodes(&self, root: ITreeNode, predicate: fn()) -> Vec<ITreeNode>;

}

/// Interface for circular reference detection.
pub trait ICircularDetector {
    /// Detect circular reference in object.
    fn detect_circular_reference(&self, obj: serde_json::Value) -> bool;

    /// Find all circular paths.
    fn find_circular_paths(&self, obj: serde_json::Value) -> Vec<Vec<serde_json::Value>>;

    /// Get objects involved in circular references.
    fn get_circular_objects(&self, obj: serde_json::Value) -> Vec<serde_json::Value>;

    /// Break circular references.
    fn break_circular_references(&self, obj: serde_json::Value) -> serde_json::Value;

}

/// Interface for graph node operations.
pub trait IGraphNode {
    /// Node ID.
    // Python decorators: @property
    fn id(&self) -> &str;

    /// Node data.
    // Python decorators: @property
    fn data(&self) -> serde_json::Value;

    /// Node neighbors.
    // Python decorators: @property
    fn neighbors(&self) -> Vec<String>;

    /// Add neighbor node.
    fn add_neighbor(&self, neighbor: String, weight: Option<f64>) -> ();

    /// Remove neighbor node.
    fn remove_neighbor(&self, neighbor: String) -> ();

    /// Get edge weight to neighbor.
    fn get_edge_weight(&self, neighbor: String) -> Option<f64>;

}

/// Interface for graph operations.
pub trait IGraph {
    /// Add node to graph.
    fn add_node(&self, node: IGraphNode) -> ();

    /// Remove node from graph.
    fn remove_node(&self, node_id: String) -> ();

    /// Get node by ID.
    fn get_node(&self, node_id: String) -> Option<IGraphNode>;

    /// Add edge between nodes.
    fn add_edge(&self, from_node: String, to_node: String, weight: Option<f64>) -> ();

    /// Remove edge between nodes.
    fn remove_edge(&self, from_node: String, to_node: String) -> ();

    /// Get all nodes.
    fn get_all_nodes(&self) -> Vec<IGraphNode>;

    /// Get all edges.
    fn get_all_edges(&self) -> Vec<()>;

    /// Check if nodes are connected.
    fn is_connected(&self, from_node: String, to_node: String) -> bool;

}
